src/mlpack/methods/reinforcement_learning/replay/prioritized_replay.hpp

```diff
@@ -0,0 +1,269 @@
+/**
+ * @file prioritized_experience_replay.hpp
+ * @author Xiaohong
+ *
+ * This file is an implementation of prioritized experience repla y.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_PRIORITIZED_REPLAY_HPP
+#define MLPACK_METHODS_RL_PRIORITIZED_REPLAY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "sumtree.hpp"
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of prioritized experience replay.
+ *
+ *
+ * @code
+ * @article{schaul2015prioritized,
+ *  title={Prioritized experience replay},
+ *  author={Schaul, Tom and Quan, John and Antonoglou, Ioannis and Silver, David},
+ *  journal={arXiv preprint arXiv:1511.05952},
+ *  year={2015}
+ *  }
+ * @endcode
+ *
+ * @tparam EnvironmentType Desired task.
+ */
+template <typename EnvironmentType>
+class PrioritizedReplay
+{
+ public:
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  /**
+   * Default constructor
+   */
+  PrioritizedReplay()
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of prioritized experience replay class.
+   *
+   * @param batchSize Number of examples returned at each sample.
+   * @param capacity Total memory size in terms of number of examples.
+   * @param alpha How much prioritization is used
+   * @param dimension The dimension of an encoded state.
+   */
+  PrioritizedReplay(const size_t batchSize,
+               const size_t capacity,
+               const double alpha,
+               int seed = 1024,
+               const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity(capacity),
+      alpha(alpha),
+      position(0),
+      states(dimension, capacity),
+      actions(capacity),
+      rewards(capacity),
+      nextStates(dimension, capacity),
+      isTerminal(capacity),
+      full(false),
+      max_priority(1.0),
+      initial_beta(0.6),
+      replay_beta_iters(10000)
+  {
+    arma::arma_rng::set_seed(seed);
+    int size = 1;
+    while (size < capacity) {
```

Comment:
Here we are making a comparison between a signed & unsigned type. Though it may not matter much as `size` will be positive, but I guess we could always avoid unnecessary warnings.

Reply:
You are right, make the type consistent when comparing is much better.